cpp/src/groupby/sort/group_single_pass_reduction_util.cuh

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/aggregation.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+
+namespace cudf {
+namespace experimental {
+namespace groupby {
+namespace detail {
+
+template <aggregation::Kind k>
+struct reduce_functor {
+
+  template <typename T>
+  static constexpr bool is_supported(){
+    if (cudf::is_numeric<T>())
+      return true;
+    else if (cudf::is_timestamp<T>() and 
+              (k == aggregation::MIN or k == aggregation::MAX))
+      return true;
+    else
+      return false;
+  }
+
+  template <typename T>
+  std::enable_if_t<is_supported<T>(), std::unique_ptr<column> >
+  operator()(column_view const& values,
+             column_view const& group_sizes,
+             rmm::device_vector<cudf::size_type> const& group_labels,
+             rmm::mr::device_memory_resource* mr,
+             cudaStream_t stream)
+  {
+    CUDF_EXPECTS(static_cast<size_t>(values.size()) == group_labels.size(),
+      "Size of values column should be same as that of group labels");
+
+    using OpType = cudf::experimental::detail::corresponding_operator_t<k>;
+    using ResultType = cudf::experimental::detail::target_type_t<T, k>;
+    size_type num_groups = group_sizes.size();
+
+    auto result = make_fixed_width_column(data_type(type_to_id<ResultType>()), 
+                                          num_groups,
+                                          values.nullable() 
+                                            ? mask_state::UNINITIALIZED
+                                            : mask_state::UNALLOCATED,
+                                          stream, mr);
+    auto op = OpType{};
+
+    if (values.size() == 0) {
+      return result;
+    }
+
+    if (values.nullable()) {
+      T default_value = OpType::template identity<T>();
+      auto device_values = column_device_view::create(values);
+      auto val_it = cudf::experimental::detail::make_null_replacement_iterator(
+                        *device_values, default_value);
+
+      // Without this transform, thrust throws a runtime error
```

Comment:
Why exactly does this fix the runtime errors you were seeing?  Because we are seeing the exact opposite in our tests. When I remove this iterator we stop seeing some crashes that we see with it in place.  Either way it feels like we really need to understand what the underlying issues are before merging this in.  Note that we saw this on an older version of this patch.  I am still trying to upmerge to this version.

Reply:
Yes I just verified that some of the crashes we were seeing go away when we remove the iterator. I have no idea at all why, though.

```diff
diff --git a/cpp/include/cudf/detail/utilities/device_operators.cuh b/cpp/include/cudf/detail/utilities/device_operators.cuh
index 62a3d7b84..978d787dc 100644
--- a/cpp/include/cudf/detail/utilities/device_operators.cuh
+++ b/cpp/include/cudf/detail/utilities/device_operators.cuh
@@ -42,8 +42,9 @@ struct DeviceSum {
   }
 
   template <typename T,
+            typename T2 = T,
             typename std::enable_if_t<!cudf::is_timestamp<T>()>* = nullptr>
-  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T2& rhs) {
     return lhs + rhs;
   }
 
diff --git a/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh b/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh
index 119760153..e1b1cf0d5 100644
--- a/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh
+++ b/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh
@@ -78,15 +78,17 @@ struct reduce_functor {
       auto val_it = cudf::experimental::detail::make_null_replacement_iterator(
                         *device_values, default_value);
 
+#if 0
       // Without this transform, thrust throws a runtime error
       auto it = thrust::make_transform_iterator(val_it,
                           [] __device__ (auto i) -> ResultType { return i; });
-      
+#endif
+
       thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
         // Input keys
           group_labels.begin(), group_labels.end(),
         // Input values
-          it,
+          val_it,
         // Output keys
           thrust::make_discard_iterator(),
         // Output values
@@ -109,7 +111,7 @@ struct reduce_functor {
         });
     } else {
       auto it = thrust::make_transform_iterator(values.data<T>(),
-                          [] __device__ (auto i) -> ResultType { return i; });
+                          [] __device__ (auto i) { return i; });
       thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
                             group_labels.begin(), group_labels.end(),
                             it, thrust::make_discard_iterator(),
```

Reply:
Do you by chance have a debug build? And did the crashes say something like `reduce_by_key failed on 2nd step: invalid device function`.

Reply:
I just removed the iterator for the non nullable case and it removed even more crashes.  I really don't understand why this is happening.

Reply:
When I remove the iterator from the nullable case, it gives me the aforementioned error in debug build for all types and in release build but only for `long`. When I remove the iterator from the nullable case, it somehow fixes the errors in release build for nullable version but debug build behaves unchanged.

Does anyone have any idea what would cause this? Is this likely a bug in my code, thrust or nvcc?

Reply:
> 
> 
> When I remove the iterator from the nullable case, it gives me the aforementioned error in debug build for all types and in release build but only for `long`. When I remove the iterator from the non-nullable case, it somehow fixes the errors in release build for nullable version but debug build behaves unchanged.
> 
> Does anyone have any idea what would cause this? Is this likely a bug in my code, thrust or nvcc?

Uhm, the problem might be in the code that follows, specifically `rmm::exec_policy(stream)->on(stream)`. I remember an old mention of _don't do this_ in transition guide (due to dangling references). 

Specifically,
``` Make sure you never do this:

           auto exec = rmm::exec_policy(stream)->on(stream);

           Else you'll be left with a reference to an invalid object.Instead:

           [...]
           auto exec = rmm::exec_policy(stream);
           thrust::sort(exec->on(stream), ...);```

Reply:
Actually, I **have** followed the advice of the post by not making the exec policy a variable and instead used it directly in the thrust call.

Reply:
I've re-written this aggregation and removed dependence on thrust in this case. Please check if this fixes the bugs.

Reply:
I retested with the latest changes and they look good.  Thanks for looking into this.